File: src/graphql/mutations/billing/quotes.ts
import { graphql } from "@/lib/graphql/client";

// ============================================================================
// QUOTE MUTATIONS
// ============================================================================

/**
 * Creates a new billing quote.
 * @param payload The input data for creating the quote.
 */
export const createQuote = graphql(`
  mutation CreateQuote($payload: CreateQuoteInput!) {
    billing {
      createQuote(payload: $payload) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the client ID of a billing quote.
 * @param id The UUID of the quote to update.
 * @param clientId The new client ID.
 */
export const updateQuoteClientId = graphql(`
  mutation UpdateQuoteClientId($id: UUID!, $clientId: UUID) {
    billing {
      updateQuoteClientId(id: $id, clientId: $clientId) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the origin details of a billing quote.
 * @param id The UUID of the quote to update.
 * @param originDetails The new origin details.
 */
export const updateQuoteOriginDetails = graphql(`
  mutation UpdateQuoteOriginDetails($id: UUID!, $originDetails: String!) {
    billing {
      updateQuoteOriginDetails(id: $id, originDetails: $originDetails) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the destination details of a billing quote.
 * @param id The UUID of the quote to update.
 * @param destinationDetails The new destination details.
 */
export const updateQuoteDestinationDetails = graphql(`
  mutation UpdateQuoteDestinationDetails($id: UUID!, $destinationDetails: String!) {
    billing {
      updateQuoteDestinationDetails(id: $id, destinationDetails: $destinationDetails) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the weight of a billing quote.
 * @param id The UUID of the quote to update.
 * @param weight The new weight.
 */
export const updateQuoteWeight = graphql(`
  mutation UpdateQuoteWeight($id: UUID!, $weight: Float) {
    billing {
      updateQuoteWeight(id: $id, weight: $weight) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the length of a billing quote.
 * @param id The UUID of the quote to update.
 * @param length The new length.
 */
export const updateQuoteLength = graphql(`
  mutation UpdateQuoteLength($id: UUID!, $length: Float) {
    billing {
      updateQuoteLength(id: $id, length: $length) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the width of a billing quote.
 * @param id The UUID of the quote to update.
 * @param width The new width.
 */
export const updateQuoteWidth = graphql(`
  mutation UpdateQuoteWidth($id: UUID!, $width: Float) {
    billing {
      updateQuoteWidth(id: $id, width: $width) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the height of a billing quote.
 * @param id The UUID of the quote to update.
 * @param height The new height.
 */
export const updateQuoteHeight = graphql(`
  mutation UpdateQuoteHeight($id: UUID!, $height: Float) {
    billing {
      updateQuoteHeight(id: $id, height: $height) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the quoted price of a billing quote.
 * @param id The UUID of the quote to update.
 * @param quotedPrice The new quoted price.
 */
export const updateQuoteQuotedPrice = graphql(`
  mutation UpdateQuoteQuotedPrice($id: UUID!, $quotedPrice: Float!) {
    billing {
      updateQuoteQuotedPrice(id: $id, quotedPrice: $quotedPrice) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the service level of a billing quote.
 * @param id The UUID of the quote to update.
 * @param serviceLevel The new service level.
 */
export const updateQuoteServiceLevel = graphql(`
  mutation UpdateQuoteServiceLevel($id: UUID!, $serviceLevel: String) {
    billing {
      updateQuoteServiceLevel(id: $id, serviceLevel: $serviceLevel) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the expiration timestamp of a billing quote.
 * @param id The UUID of the quote to update.
 * @param expiresAt The new expiration timestamp.
 */
export const updateQuoteExpiresAt = graphql(`
  mutation UpdateQuoteExpiresAt($id: UUID!, $expiresAt: DateTime) {
    billing {
      updateQuoteExpiresAt(id: $id, expiresAt: $expiresAt) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the status of a billing quote.
 * @param id The UUID of the quote to update.
 * @param status The new status.
 */
export const updateQuoteStatus = graphql(`
  mutation UpdateQuoteStatus($id: UUID!, $status: QuoteStatusEnum!) {
    billing {
      updateQuoteStatus(id: $id, status: $status) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the quote number of a billing quote.
 * @param id The UUID of the quote to update.
 * @param quoteNumber The new quote number.
 */
export const updateQuoteQuoteNumber = graphql(`
  mutation UpdateQuoteQuoteNumber($id: UUID!, $quoteNumber: String) {
    billing {
      updateQuoteQuoteNumber(id: $id, quoteNumber: $quoteNumber) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the notes of a billing quote.
 * @param id The UUID of the quote to update.
 * @param notes The new notes.
 */
export const updateQuoteNotes = graphql(`
  mutation UpdateQuoteNotes($id: UUID!, $notes: String) {
    billing {
      updateQuoteNotes(id: $id, notes: $notes) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the created by user ID of a billing quote.
 * @param id The UUID of the quote to update.
 * @param createdByUserId The new created by user ID.
 */
export const updateQuoteCreatedByUserId = graphql(`
  mutation UpdateQuoteCreatedByUserId($id: UUID!, $createdByUserId: UUID) {
    billing {
      updateQuoteCreatedByUserId(id: $id, createdByUserId: $createdByUserId) {
        id
        originDetails
        destinationDetails
        weight
        length
        width
        height
        volume
        quotedPrice
        serviceLevel
        expiresAt
        status
        quoteNumber
        notes
        createdAt
        updatedAt
        client {
          id
          name
        }
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Removes a billing quote by its ID.
 * @param id The UUID of the quote to remove.
 */
export const removeQuote = graphql(`
  mutation RemoveQuote($id: UUID!) {
    billing {
      removeQuote(id: $id)
    }
  }
`);

File: services/graphql-billing/src/mutation/quotes.rs
use async_graphql::{Context, InputObject, Object};
use chrono::{DateTime, Utc};
use uuid::Uuid;

use crate::models::{enums::QuoteStatusEnum, quotes};

#[derive(Debug, Clone, InputObject)]
pub struct CreateQuoteInput {
    pub client_id: Option<Uuid>,
    pub origin_details: String,
    pub destination_details: String,
    pub weight: Option<f64>,
    pub length: Option<f64>,
    pub width: Option<f64>,
    pub height: Option<f64>,
    pub quoted_price: f64,
    pub service_level: Option<String>,
    pub expires_at: Option<DateTime<Utc>>,
    pub status: Option<QuoteStatusEnum>,
    pub quote_number: Option<String>,
    pub notes: Option<String>,
    pub created_by_user_id: Option<Uuid>,
}

#[derive(Debug, Clone, Default)]
pub struct QuotesMutation;

#[Object(name = "BillingQuotesMutation")]
impl QuotesMutation {
    async fn create_quote(
        &self,
        ctx: &Context<'_>,
        payload: CreateQuoteInput,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "insert into billing.quotes (client_id, origin_details, destination_details, weight, length, width, height, quoted_price, service_level, expires_at, status, quote_number, notes, created_by_user_id) values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14) returning *",
        )
        .bind(payload.client_id)
        .bind(payload.origin_details)
        .bind(payload.destination_details)
        .bind(payload.weight)
        .bind(payload.length)
        .bind(payload.width)
        .bind(payload.height)
        .bind(payload.quoted_price)
        .bind(payload.service_level)
        .bind(payload.expires_at)
        .bind(payload.status)
        .bind(payload.quote_number)
        .bind(payload.notes)
        .bind(payload.created_by_user_id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_client_id(
        &self,
        ctx: &Context<'_>,
        client_id: Option<Uuid>,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set client_id = $1 where id = $2 returning *",
        )
        .bind(client_id)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_origin_details(
        &self,
        ctx: &Context<'_>,
        origin_details: String,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set origin_details = $1 where id = $2 returning *",
        )
        .bind(origin_details)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_destination_details(
        &self,
        ctx: &Context<'_>,
        destination_details: String,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set destination_details = $1 where id = $2 returning *",
        )
        .bind(destination_details)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_weight(
        &self,
        ctx: &Context<'_>,
        weight: Option<f64>,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set weight = $1 where id = $2 returning *",
        )
        .bind(weight)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_length(
        &self,
        ctx: &Context<'_>,
        length: Option<f64>,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set length = $1 where id = $2 returning *",
        )
        .bind(length)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_width(
        &self,
        ctx: &Context<'_>,
        width: Option<f64>,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set width = $1 where id = $2 returning *",
        )
        .bind(width)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_height(
        &self,
        ctx: &Context<'_>,
        height: Option<f64>,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set height = $1 where id = $2 returning *",
        )
        .bind(height)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_quoted_price(
        &self,
        ctx: &Context<'_>,
        quoted_price: f64,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set quoted_price = $1 where id = $2 returning *",
        )
        .bind(quoted_price)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_service_level(
        &self,
        ctx: &Context<'_>,
        service_level: Option<String>,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set service_level = $1 where id = $2 returning *",
        )
        .bind(service_level)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_expires_at(
        &self,
        ctx: &Context<'_>,
        expires_at: Option<DateTime<Utc>>,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set expires_at = $1 where id = $2 returning *",
        )
        .bind(expires_at)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_status(
        &self,
        ctx: &Context<'_>,
        status: QuoteStatusEnum,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set status = $1 where id = $2 returning *",
        )
        .bind(status)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_quote_number(
        &self,
        ctx: &Context<'_>,
        quote_number: Option<String>,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set quote_number = $1 where id = $2 returning *",
        )
        .bind(quote_number)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_notes(
        &self,
        ctx: &Context<'_>,
        notes: Option<String>,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set notes = $1 where id = $2 returning *",
        )
        .bind(notes)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_quote_created_by_user_id(
        &self,
        ctx: &Context<'_>,
        created_by_user_id: Option<Uuid>,
        id: Uuid,
    ) -> async_graphql::Result<quotes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, quotes::Model>(
            "update billing.quotes set created_by_user_id = $1 where id = $2 returning *",
        )
        .bind(created_by_user_id)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn remove_quote(&self, ctx: &Context<'_>, id: Uuid) -> async_graphql::Result<String> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query("delete from billing.quotes where id = $1")
            .bind(id)
            .execute(db)
            .await?;

        if result.rows_affected() != 1 {
            return Err(async_graphql::Error::new("Unable to remove quote"));
        }

        Ok("Quote removed successfully".into())
    }
}
